File: AmaroTest/app/src/main/java/com/test/amaro/amarotest/dagger/modules/ApplicationModule.java
package com.test.amaro.amarotest.dagger.modules;

import android.content.Context;
import android.content.SharedPreferences;
import com.test.amaro.amarotest.AmaroApplication;
import com.test.amaro.amarotest.dagger.scopes.ApplicationContext;
import dagger.Module;
import dagger.Provides;
import javax.inject.Singleton;

@Module
public final class ApplicationModule {

    private final AmaroApplication androidApplication;
    private static final String PREFERENCE = "preference";

    @Provides
    @Singleton
    public final AmaroApplication provideApplication() {
        return this.androidApplication;
    }

    @Provides
    @Singleton
    @ApplicationContext
    public final Context provideApplicationContext() {
        return this.androidApplication;
    }

    public ApplicationModule(AmaroApplication androidApplication) {
        this.androidApplication = androidApplication;
    }

    @Provides
    @Singleton
    public SharedPreferences provideSharedPreferences() {
        return androidApplication.getSharedPreferences(PREFERENCE, Context.MODE_PRIVATE);
    }
}

File: AmaroTest/app/src/main/java/com/test/amaro/amarotest/dagger/modules/ApplicationModule.kt
package com.test.amaro.amarotest.dagger.modules

import android.content.Context
import com.test.amaro.amarotest.dagger.scopes.ApplicationContext
import com.test.amaro.amarotest.AmaroApplication
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

/**
 * Dagger module that provides objects which will live during the application lifecycle.
 */
@Module
class ApplicationModule constructor(private val androidApplication: AmaroApplication) {

    @Provides
    @Singleton
    internal fun provideApplication(): AmaroApplication {
        return androidApplication
    }

    @Provides
    @Singleton
    @ApplicationContext
    internal fun provideApplicationContext(): Context {
        return androidApplication
    }

}
